### DISCUSSION TITLE
A virtualized table implementation with per cell stale time

### DISCUSSION THREAD

**Discussion by marbemac** [Omnistac/zedux#134] [Show and tell]
Hi @bowheart - now that the holidays are over, I finally had a chance to dig into Zedux with a more complex example. This is that virtualized data table / grid that I mentioned earlier.

I had some fun putting it together today, BUT I'm not sure that I did things the "zedux" way across the board. It's easy to run locally, and I included a quick gif in the readme to give an idea of what it does. I also added some comments tagged with `@QUESTION` with notes on bits I was particularly unsure about as I put it together.

No rush at all, but if you have a little time in the next couple of weeks and are able to take a look at the implementation to offer some pointers I'd really appreciate it. Also, if a more advanced example like this is something that could be helpful to others I could clean it up (and/or simplify it) for broader consumption.

Thanks in advance!

https://github.com/marbemac/zedux-grid-example

**Comment by bowheart** [comment]
@marbemac amazing! I'll try to find some time to look over it in the next few days.

**Reply by marbemac** [reply]
🙏   thanks - could also update to use specific features (like the DI thing mentioned in https://github.com/Omnistac/zedux/discussions/135#discussioncomment-11454822), and then call out those patterns in the readme or something, with links to the in-code example implementation. As a showcase for these more advanced patterns.

**Reply by david-trumid** [reply]
@marbemac did notice in your atoms that you are using `injectCallback` on functions that you are also adding on exports. So for example the params to the atom `{ objectId }` and the dep list to the inject callbacks are ` [objectId]`, Since the objectId is part of the atom it makes the use of injectCallback not needed and you can remove it. Also be aware that exports will NOT update once set so even if it did update the export would have the original function on it

**Reply by marbemac** [reply]
@david-trumid thanks for taking a look! Hmm, so when WOULD one use injectCallback? I thought we needed to use it in order to give the exports api a stable reference, as in this example -> https://omnistac.github.io/zedux/docs/walkthrough/atom-apis#exporting-callbacks

I have since learned that yeah atom params don't need to be in deps lists, so will remove that 👍 .

**Reply by david-trumid** [reply]
right.. in that example notice there is a "tip" 

```
injectCallback is the injector equivalent of useCallback. We used it here with empty deps ([]) to remind us that this is an export and its reference can't change. This means we can't use any dynamic values inside the callback - only stable references. injectCallback also automatically [batches](https://omnistac.github.io/zedux/docs/advanced/batching) all state updates in the callback.
```

so thats sort of saying reminder that its not going to change, however any time the atom reevaluates it will still rebuild that function but the dep list will just ignore it .. having the same net result.

If your using an inject effect however you could use the callback as a dependency inside of that for instance.

Ideally we have found that limiting the atom itself from rerunning helps alot (such as not subscribing to internal stores, or using getInstance etc.)

**Reply by marbemac** [reply]
Ah I think I understand now. Is this right:

```ts
/**
 * Stable ref
 */
const myAtom = atom('my-atom', () => {
  const toggleUser = injectCallback(() => {}, [])

  return api(store).setExports({ toggleUser })
})

/**
 * NOT a stable ref, BUT does not matter since `setExports` will not update
 * on atom runs after the first run
 */
const myAtom = atom('my-atom', () => {
  const toggleUser = () => {};

  return api(store).setExports({ toggleUser })
})
```

So the reason it works without `injectCallback` is because of the setExports behavior, right?

> injectCallback is the injector equivalent of useCallback

Yeah I saw that and that is what actually made me think I needed to use injectCallback to provide a stable reference to the exports api.

A lot of the docs draws comparisons between atoms, their injectors, and react equivalents. So I figured passing a function to setExports would be kind of like passing a func to a child component in react, and that was the reason the example used `injectCallback` with an empty deps array.

Because in react the usage of useCallback with empty deps can actually make a difference. So that's I think what confused me in this case.

```tsx
const MyComp = () => {
  const toggleUser = useCallback(() => {}, []);

  return <SomeMemoWrappedComp onClick={toggleUser} />;
}

const MyComp = () => {
  /** not stable, actually matters in react if care about re-renders **/
  const toggleUser = () => {};

  return <SomeMemoWrappedComp onClick={toggleUser} />;
}
```

**Reply by david-trumid** [reply]
yup! exactly and totally understand, using the inject callback there def isn't wrong, we used to do it originally as well. For simple atoms alot of times we inline the function right in the export setter itself ..
```
const myAtom = atom('my-atom', () => {
  return api(store).setExports({ 
	toggleUser: () => {};
  })
})
```

**Reply by marbemac** [reply]
Hmm ok a couple of follow-up questions. 

I've got a `populate` function using `injectCallback` here -> https://github.com/marbemac/zedux-grid-example/blob/main/app/atoms/records.ts#L165

1. I'm passing `store` into the deps array, is that unnecessary since if I remember correctly `injectStore` returns a stable ref anyways?
2. This `populate` function is exported in the atom's api. Is this a problem, since the `records` value from `injectAtomInstance` is passed into the `populate` deps? Is it not possible to export functions that use dynamic values (via deps array)?

So far I think my difficulties with Zedux can be broadly broken down into two categories:

1. There sometimes seem to be multiple ways to accomplish the same thing, which causes me to be unsure about when to use what and if there are tradeoffs. An example of this would be atom getters vs injectors (e.g. `getInstance` from `injectAtomGetters` vs `injectAtomInstance`).
2. The "rules of zedux" - similar to react's "rules of hooks" except it seems that zedux has even more of these tricky behaviors (like remembering that setExports only actually does something on the atom's first run). A little worried about myself (or other devs on the team) introducing subtle bugs because they don't know about or remember these rules all the time.

Enjoying Zedux and I like a lot of the ideas in it! Just thought I'd share some initial feedback as a new user.

**Reply by david-trumid** [reply]
Love the feedback!, so in this case 

```
    const populate = injectCallback(
      (force = false) => {
        const { value, version, populatedAt } = store.getState();

        /**
         * Consider the value stale if it was populated more than X seconds ago, and refetch it.
         */
        const isStale =
          populatedAt && Date.now() - STALE_DATA_THRESHOLD > populatedAt;

        if (!force && value !== undefined && !isStale) {
          return;
        }

        records.exports.populateAttribute(recordId, columnId);
      },
      [store, records]
    );
```

both `store` (a Store) and `records` (an atom instance) are stable references and neither are technically needed in this depList. 

and since it is only using stable references you dont need the injectCallback and it is perfectly safe to export it. Basically as long as your exports are accessing stable references you are good. You are good to use the dynamic values because you are accessing them from the export itself which is fine.

On the multiple ways..

Totally understood, think we will be simplifying these in future versions (@bowheart can confirm for sure) I could go into background of how it evolved if you would like but a quick summary is those things are basically the same thing under the covers just useful in different scenarios. For instance if all you need is to inject and atom instance then all you need is `injectAtomInstance`.. however if you need to inject values, instances, selectors or multiple of each.. there is a helper to expose all of them. Not to confuse you more but.. if you are deriving an atom from other atoms (like your `recordAttributeAtom` is based off `recordsAtom`) then you might enjoy using `ion`'s for those, those give you the atom getters as the first parameter and resemble more like recoil. For instance..

```
export const recordAttributeAtom = ion(
 { getInstance },
  "record-attribute",
  ({
    objectId,
    recordId,
    columnId,
  }: {
    objectId: string;
    recordId: string;
    columnId: string;
  }) => {
    const store = injectStore({
      value: undefined as string | undefined,
      version: 0, // Tracking this just to show in the UI when a record attribute is re-fetched, for example purposes
      populatedAt: null as number | null,
    });

    const records = getInstance(recordsAtom, [{ objectId }]);
```

It makes more sense as you get more and more complicated and need multiple instances/values/selectors.

**Reply by marbemac** [reply]
> and since it is only using stable references you dont need the injectCallback and it is perfectly safe to export it

Cool that's what I figured. Atm is it not possible to expose functions with dynamic deps via the api exports?

> Totally understood, think we will be simplifying these in future versions

❤️ 

> Not to confuse you more but.. if you are deriving an atom from other atoms (like your recordAttributeAtom is based off recordsAtom) then you might enjoy using ion's for those

Is deriving atoms from other atoms a common use case, or should I try to avoid that somehow? Asking because I'm doing this all over the place ha.

Hmm, hadn't even considered ions. Haven't looked much into it, on initial quick read of the docs I categorized it in my head as "ok this is for when I need more complex computed data memoized" - is that indeed the main use case? In this example does it provide any benefit beyond exposing the atom getter functions?

Thanks for your help!

**Reply by david-trumid** [reply]
No Ion is just some syntactic sugar :D 

We have thousands of atoms and we have yet to need an export to be dynamic (since the consumer calls it and it will then have access to anything in the atom definition which can then be stable as well) so I think we haven't actually needed that use case. Usually if the consumer wants to call a dynamic function they would be subscribing to the atom store itself and when it calls the export the export can access through stable references

**Reply by bowheart** [reply]
> is it not possible to expose functions with dynamic deps via the api exports?

Atoms support this via state. If the function ref needs to change, then it's state. Consumers need to be notified of the change. Stores handle all that. There's nothing wrong with putting functions in state (as long as you're not trying to serialize them).

Exports are different. Treat them like JS module exports - when you `export function MyExport() {}`, that function reference never changes.

We do have multiple ways to do things, and yeah there's a history behind all of it. We can discuss cleaning some of those up. In fact, I just opened #143 to discuss utils packages. Some of Zedux's core APIs should have probably been utils in there.

> Is deriving atoms from other atoms a common use case

This is what atoms are for. It's why atomic libraries scale better than singletons like Redux and is probably the main reason why Zedux exists.

**Comment by bowheart** [comment]
Updating here for completeness: I did a deep dive on this project, answering all questions and migrating the code to use Zedux v2. Amazing project @marbemac thanks for sharing.

See my fork PR: https://github.com/marbemac/zedux-grid-example/pull/1
